### DISCUSSION TITLE
Why when watching an array the autorun will run two times?

### DISCUSSION THREAD

**Discussion by joaovct** [mobxjs/mobx#3169] [Q&A] [answered]
Taking the following [sandbox](https://codesandbox.io/s/autorun-pdvq1?file=/src/App.js), my question is very simple: why the `autorun` callback is being called two times for each update of the `apples` array? (Check the console.log for seeing it).
I also tested with `reaction` and passing `apples` as dependence seems to work good. But just want to clarify the behavior of `autorun` on this one.

**Comment by urugator** [comment, marked as answer]
`useEffect` without depenency list runs on every render - so you recreate `autorun` every render. Since `observer` observes `apples.length` it's re-rendered on every `apples` change - this means one `useEffect` and therefore one `autorun` invocation per render.
Additionally the `autorun` also observes `apples.length` and it's actually invoked before the `observer` (the order is not defined).
So one is the `autorun` being notified directly, the second is recreation of the autorun by invalidated `observer`.
See https://github.com/mobxjs/mobx/discussions/3058#discussioncomment-1162656

**Reply by joaovct** [reply]
Thanks! Got it working by passing `apples` as a `useEffect` dependency. So, I guess the short-form answer here is for not creating the `autorun` on every re-render?
 By the way, because I'm tied up, I've only skimmed a few issues related to the topic of reactions (such as [#772](https://github.com/mobxjs/mobx-react/issues/772), [#853](https://github.com/mobxjs/mobx-react/issues/853) and [#1006](https://github.com/mobxjs/mobx/issues/1006)). I'm not sure about not using it at all, and go entirely for imperative. Yet, I won't hit the hammer, because I'm not fully aware of everything about it. Anyway, thank you!
